Implementação de assinaturas híbridas com a liboqs

alexandreboutrik opened the pull request:
**Descrição**:
Implementa um sistema de assinaturas híbridas para o projeto [BBPQ](https://pqc-group-utfpr.github.io/projects.html).
@Yuuto-Senpai ficaria encarregado de realizar benchmarks e testes.

**Uso geral**:
1. `GenerateKey`: função para criar uma chave privada;
2. `PrivateKey.ExportPublicKey`: método para extrair a chave pública;
3. `PrivateKey.GetPrivateKeys`: método para obter as chaves públicas clássica e PQC;
5. `PrivateKey.Sign`: método para assinar uma mensagem;
6. `PublicKey.GetPublicKeys`: método para obter as chaves públicas clássica e PQC;
7. `Verify`: função para verificar a validade de uma assinatura híbrida, clássica ou PQC.

**Observações**:
- Segui o modelo de struct utilizado em `crypto/ecdsa` e `crypto/rsa`, onde a struct `PrivateKey` encapsula métodos e mantém uma referência para a struct `PublicKey`. Como a função `GenerateKey` retorna apenas a `PrivateKey`, é necessário utilizar um método `PrivateKey.Export*` para acessar a `PublicKey`.
- A assinatura híbrida é a concatenação das assinaturas clássica e PQC, e ela **não** faz parte das structs, sendo somente passada como argumento e retornada pelos métodos e/ou funções. Não sei se seria interessante armazená-la, criando um novo membro `hybrid []byte` na `PublicKey` ou `PrivateKey`;
- Por estar na codebase da go-std, os commits tentam seguir as convenções definidas pelo CONTRIBUTING.md ao invés da Conventional Commits 1.0.0.

Closes: #1

alexandreboutrik committed:
crypto/hybrid: add hybrid signatures

This commit introduces the implementation of hybrid signature cryptography for the BBPQ project, developed by the Post-Quantum Cryptography Group at UTFPR-TD.

The following signature schemes are supported:
- MAYO1_P256
- MAYO2_P256
- MAYO3_P384
- MAYO5_P521

Implemented functions:
- `GenerateKey`: Generates an hybrid key pair.
- `Sign`: creates signatures using an hybrid private key.

Note: The `Verify` function is not yet implemented and remains as a
TODO.

alexandreboutrik committed:
crypto/hybrid: add Verify

Add `Verify` function.

Changes:
- `PrivateKey.ExportPublicKey` method now returns a `PublicKey` struct instead of directly returning the classic and PQC public keys.
- To retrieve the individual public keys (classic and PQC), use the `PublicKey.GetPublicKeys` method.

With this commit, the implementation should now be feature-complete.

alexandreboutrik committed:
crypto/hybrid: change order of the sigs in hybridSig

Classic signatures are usually smaller in size (in bytes) compared to PQC signatures. Therefore, concatenating PQC signatures to Classic signatures is likely to be more efficient.

PQC-Group-UTFPR reviewed (changes requested):
Oi @alexandreboutrik e @Yuuto-Senpai,

segue comentários:
- Sobre a estrutura, eu acho que vale manter desta forma, só peço que deem uma pesquisada sobre as funções de `Marshal` e `Unmarshal` pois talvez precisaremos serializar de alguma forma no futuro.
- Eu só vou pedir uma modificação antes de aprovar, que é em relação aos [comentários imediatamente](https://tip.golang.org/doc/comment) acima das funções, const, pacotes, etc. Em especial comentar que a função `SignASN1` está esperando o hash da mensagem e não a mensagem propriamente dita (correto?). Mesma coisa para a parte PQC.
- E já que pedi modificação para especificar isso, aproveitem para estabelecer a política do `Verify`: verificar ambas obrigatoriamente (i.e., remover o `switch option`).

gabrielhbueno commented:
Ok.

Em qua., 27 de nov. de 2024, 10:18, Grupo PQC UTFPR-TD <
***@***.***> escreveu:

> ***@***.**** requested changes on this pull request.
>
> Oi @alexandreboutrik <https://github.com/alexandreboutrik> e @Yuuto-Senpai
> <https://github.com/Yuuto-Senpai>,
>
> segue comentários:
>
>    - Sobre a estrutura, eu acho que vale manter desta forma, só peço que
>    deem uma pesquisada sobre as funções de Marshal e Unmarshal pois
>    talvez precisaremos serializar de alguma forma no futuro.
>    - Eu só vou pedir uma modificação antes de aprovar, que é em relação
>    aos comentários imediatamente <https://tip.golang.org/doc/comment>
>    acima das funções, const, pacotes, etc. Em especial comentar que a função
>    SignASN1 está esperando o hash da mensagem e não a mensagem
>    propriamente dita (correto?). Mesma coisa para a parte PQC.
>    - E já que pedi modificação para especificar isso, aproveitem para
>    estabelecer a política do Verify: verificar ambas obrigatoriamente
>    (i.e., remover o switch option).
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/PQC-Group-UTFPR/go-std/pull/2#pullrequestreview-2465002200>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AUSW5WS6ODEP4TFYFCGLRS32CXBDNAVCNFSM6AAAAABSA5BCIKVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDINRVGAYDEMRQGA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

PQC-Group-UTFPR commented:
Aproveitando; alguma chance da estrutura incluir edDSA? Ao invés do ECDSA. Ou criar uma nova. Seria interessante em alguns testes ter o mapeamento do nível um com ed25519 ao invés da P-256.

alexandreboutrik committed:
crypto/hybrid: Add Doc comments, refactor Verify fn

Changes:
- `Verify` function renamed to `VerifyHybrid`.
- Doc comments added to all public functions and methods.
- Clarify (by Doc comments) that the functions expects the signed hash of a message, not the raw message itself (hashing is not performed on the provided arguments).

TODO:
- `Marshal` and `Unmarshal`.

alexandreboutrik committed:
crypto/hybrid: added CROSS and ED25519.

Updated `SigName.classic` to be of type `interface{}` to support polymorphism, allowing it to hold either an `elliptic.Curve` or an `ed25519.PrivateKey` as needed.

Changes:
- Added `CROSS ED25519` according to issue #4.
- Added `MAYO ED25519`.

alexandreboutrik commented:
https://github.com/PQC-Group-UTFPR/go-std/pull/2#issuecomment-2508479868 :
> Aproveitando; alguma chance da estrutura incluir edDSA? Ao invés do ECDSA. Ou criar uma nova. Seria interessante em alguns testes ter o mapeamento do nível um com https://github.com/PQC-Group-UTFPR/go-std/commit/ed2551978a9c5037027f96f1658aee307fd5c026 ao invés da P-256.

Boa tarde @PQC-Group-UTFPR 

Acredito que sim, podemos utilizar polimorfismo com interface{} e type switches, como fiz no commit https://github.com/PQC-Group-UTFPR/go-std/pull/2/commits/f983421c6dd3bd9bfa60613c1fd14555a08b6d29 - porém não tenho certeza se essa seria a forma mais eficiente de se abordar isso.
Essa opção ofereceria uma camada de abstração para o usuário (não vai precisar se preocupar se está utilizando `ecdsa`, `ed25519`, e etc). Porém, creio que sacrificaria um pouco de desempenho (não sei como isso vai impactar nos benchmarks).

Talvez seja melhor criar structs dedicadas para `ecdsa` e `ed25519`?

PQC-Group-UTFPR commented:
Sobre o uso de interfaces, Creio que não vai impactar nos benchmarks, talvez um profiling futuro nos dê mais informações. 

Creio que faltou o CROSS com as curvas do NIST (P-256, P-384, P-521) para os parâmetros `small`. Pode adicionar? E aí fechamos a issue.

Obrigado

alexandreboutrik committed:
crypto/hybrid: add CROSS ECDSA/NIST maps.

bifee committed:
crypto/hybrid: added ML-DSA with ECDSA and ED25519.

Updated 'GetPublicKeys' and 'GetPrivateKeys' return type to 'interface{}' to improve polyforphism.

Addes support for the following signatures schemes:
- ML_DSA_44_P256
- ML_DSA_65_P384
- ML_DSA_87_P521
- ML_DSA_65_ED25519

bifee commented:
Adicionado suporte ao algoritmo ML-DSA com ECDSA e ED25519(43f1409):
- ML_DSA_44_P256 (lvl 2)
- ML_DSA_65_P384 (lvl 3)
- ML_DSA_87_P521 (lvl 5)
- ML_DSA_65_ED25519 (lvl 3)

Alterado o tipo de retorno de 'GetPublicKeys' e 'GetPrivateKeys' para 'interface{}' para melhorar o polimorfismo.

Closes: #5

PQC-Group-UTFPR commented:
Pessoal, uma dúvida:

por que `ed25519.PrivateKey{}` em `var SigOIDtoName = map[OID]SigName{...` ? Se quiser poderia inserir uma string de definição nossa. Se puderem verificar (ou trocar) em todos as ocorrências @alexandreboutrik @bifee

alexandreboutrik committed:
crypto/hybrid: refactor map

Due to potential security concerns, the `sigOIDtoName` map has been made
private (now starting with a lowercase letter), and the classic key type
has been updated from `Ed25519.PrivateKey` to `Ed25519.PublicKey`.
